File: challenge-2/challenge-2.1/challenge-2.1.js
class BankAccount {
    constructor(accountNumber, accountHolderName, initialBalance) {
        this._accountNumber = accountNumber;
        this._accountHolderName = accountHolderName;
        this._accountBalance = initialBalance || 0;
    }

    // Method to deposit money into the account
    deposit(amount) {
        if (amount > 0) {
            this._accountBalance += amount;
            console.log(`Deposited $${amount}. New balance: $${this._accountBalance}`);
        } else {
            console.log("Invalid deposit amount. Amount must be greater than 0.");
        }
    }

    // Method to withdraw money from the account
    withdraw(amount) {
        if (amount > 0 && amount <= this._accountBalance) {
            this._accountBalance -= amount;
            console.log(`Withdrawn $${amount}. New balance: $${this._accountBalance}`);
        } else if (amount > this._accountBalance) {
            console.log("Insufficient funds.");
        } else {
            console.log("Invalid withdrawal amount. Amount must be greater than 0.");
        }
    }

    // Method to display the account balance
    displayBalance() {
        console.log(`Account balance for ${this._accountHolderName}: $${this._accountBalance}`);
    }
}

// Create an instance of the BankAccount class
const myAccount = new BankAccount("12345", "John Doe", 1000);

// Test deposit and withdrawal functionality
myAccount.displayBalance(); // Display initial balance
myAccount.deposit(500);     // Deposit $500
myAccount.withdraw(200);    // Withdraw $200
myAccount.withdraw(1500);   // Attempt to withdraw more than the balance
myAccount.withdraw(-50);    // Attempt to withdraw a negative amount
myAccount.displayBalance(); // Display updated balance

File: challenge-2/challenge-2.1/challenge-2.1.py
class BankAccount:
    def __init__(self, account_number, account_holder_name, initial_balance=0):
        self.__account_number = account_number
        self.__account_holder_name = account_holder_name
        self.__account_balance = initial_balance

    def deposit(self, amount):
        if amount > 0:
            self.__account_balance += amount
            return f"Deposited ${amount}. New balance: ${self.__account_balance}"
        else:
            return "Invalid deposit amount. Amount must be greater than zero."

    def withdraw(self, amount):
        if amount > 0 and amount <= self.__account_balance:
            self.__account_balance -= amount
            return f"Withdrew ${amount}. New balance: ${self.__account_balance}"
        elif amount > self.__account_balance:
            return "Insufficient funds."
        else:
            return "Invalid withdrawal amount. Amount must be greater than zero."

    def display_balance(self):
        return f"Account Balance for {self.__account_holder_name}: ${self.__account_balance}"


# Testing the BankAccount class
if __name__ == "__main__":
    # Creating an instance of the BankAccount class
    account1 = BankAccount("123456789", "John Doe", 1000)

    # Depositing and withdrawing money
    print(account1.deposit(500))
    print(account1.withdraw(200))
    print(account1.display_balance())

    # Attempting an invalid withdrawal
    print(account1.withdraw(1500))

